### PR TITLE
fix: Fail on errors during initial setup

### LINKED ISSUE #869: Permission error is not showing up in console log
### Describe the bug

Running cloudquery with the following account configuration which contains incorrect assume role or doesn't have permissions will just quiet:

```
    accounts "test" {
      // Optional. Role ARN we want to assume when accessing this account
      role_arn = "arn:aws:iam::615713231484:role/OrganizationAccountAccessRole"
      // Optional. Local Profile is the named profile in your shared configuration file (usually `~/.aws/config`) that you want to use for this specific account
      //local_profile = "<NAMED_PROFILE>
      // Optional. Specify the Role Session name
      //role_session_name = ""
    }
```

and if and only if you run it with `--enable-console-log` it will show multiple error 

```
8:51PM ERR error retrieving credentials @module=aws err="failed to refresh cached credentials, operation error STS: AssumeRole, https response error StatusCode: 403, RequestID: eb4d5fbe-2f9a-43da-8b20-bd78xxxxxx, api error AccessDenied: User: arn:aws:sts::613483666403:assumed-role/AWSReservedSSO_AdministratorAccess_6de2xxxxxxxx/yp@cloudquery.io is not authorized to perform: sts:AssumeRole on resource: arn:aws:iam::613483666403:role/aws-service-role/organizations.amazonaws.com/AWSServiceRoleForOrganizations" timestamp=2022-05-09T20:51:53.798+0300
8:51PM INF provider configured successfully alias=aws diags="STS: AssumeRole - User: arn:aws:sts::6134xxxxx:assumed-role/AWSReservedSSO_AdministratorAccess_6xxxxxx/yp@cloudquery.io is not authorized to perform: sts:AssumeRole on resource: arn:aws:iam::61348xxxxx:role/aws-service-role/organizations.amazonaws.com/AWSServiceRoleForOrganizations: You are not authorized to perform this operation. Check your IAM policies, and ensure that you are using the correct access keys." provider=aws version=v0.11.3
8:51PM INF requesting provider fetch alias=aws provider=aws version=v0.11.3
8:51PM INF provider started fetching resources alias=aws provider=aws version=v0.11.3
8:51PM ERR received provider fetch error error="rpc error: code = Unknown desc = provider client is not configured (Hint: Try upgrading cloudquery)" alias=aws provider=aws version=v0.11.3
```

### Expected Behavior

It should print the permission error and diag to the console

### Steps to Reproduce

cloudquery fetch

### Possible Solution

_No response_

### Provider and CloudQuery version

0.11.3

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by erezrokah**
🎉 Thank you for making CloudQuery awesome by submitting a PR 🎉

#### Summary

Fixes https://github.com/cloudquery/cq-provider-aws/issues/869

---

Use the following steps to ensure your PR is ready to be reviewed

- [x] Read the [contribution guidelines](../blob/main/CONTRIBUTING.md) 🧑‍🎓
- [x] Run `go fmt` to format your code 🖊
- [x] Update or add tests. Learn more about testing [here](https://docs.cloudquery.io/docs/developers/sdk/testing) 🧪
- [x] Update the docs by running `go run ./docs/docs.go` and committing the changes 📃
- [x] If adding a new resource, add [relevant Terraform files](../terraform) in a separate PR 📂
- [x] Ensure the status checks below are successful ✅

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -442,7 +442,7 @@ func configureAwsClient(ctx context.Context, logger hclog.Logger, awsConfig *Con
 	// Test out retrieving credentials
 	if _, err := awsCfg.Credentials.Retrieve(ctx); err != nil {
 		logger.Error("error retrieving credentials", "err", err)
-		return awsCfg, fmt.Errorf(awsFailedToConfigureErrMsg, account.AccountName, err, checkEnvVariables())
+		return awsCfg, diag.FromError(fmt.Errorf(awsFailedToConfigureErrMsg, account.AccountName, err, checkEnvVariables()), diag.USER)
```
Comment (erezrokah) [inline client/client.go:445]: This fixes a similar issue to https://github.com/cloudquery/cq-provider-aws/issues/869.
Happened to me on my default AWS profile which is my personal account. The default account failed to assume the role, so it reached this code.

File: client/client.go [hunk 2]
```diff
@@ -515,7 +515,7 @@ func Configure(logger hclog.Logger, providerConfig interface{}) (schema.ClientMe
 				}
 			})
 		if err != nil {
-			return nil, diags.Add(classifyError(fmt.Errorf("failed to find disabled regions for account %s. AWS Error: %w", account.AccountName, err), diag.INTERNAL, nil))
+			return nil, diags.Add(diag.FromError(fmt.Errorf("failed to find disabled regions for account %s. AWS Error: %w", account.AccountName, err), diag.USER))
```
[the remark this answers is not in the document]
Comment (disq) [reply client/client.go:518]: This misses the point of `classifyError` and forces every error to be diag.USER.

[the remark this answers is not in the document]
Comment (erezrokah) [reply client/client.go:518]: Good point. But we do want to report authentication errors as user errors in this specific case (during the setup phase before fetch).

File: client/client.go (hunk 1 above)
Comment (disq) [reply client/client.go:445]: again, this is called by `client.Configure` which calls classifyError on the result. As long as the error is wrapped with %w the classification should be done in `classifyError`.

Comment (erezrokah) [reply client/client.go:445]: Would it make sense to separate auth errors during the setup phase from auth errors during the actual resources resolving phase?

If we failed outright at the beginning with an auth error can we assume it's a user error?

File: client/client.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (disq) [reply client/client.go:518]: Yes but how do you know it's an authentication error without classifying it? If you think all errors from `Configure` should be classified as such, replace all calls to `classifyError` with hard USER errors, and not just a couple.

[the remark this answers is not in the document]
Comment (erezrokah) [reply client/client.go:518]: Ah, I get it now. Let me see if I can scope the code to auth errors

File: client/client.go [hunk 3]
```diff
@@ -442,6 +442,10 @@ func configureAwsClient(ctx context.Context, logger hclog.Logger, awsConfig *Con
 	// Test out retrieving credentials
 	if _, err := awsCfg.Credentials.Retrieve(ctx); err != nil {
 		logger.Error("error retrieving credentials", "err", err)
+		classifiedError := classifyError(err, diag.INTERNAL, nil)
+		if classifiedError[0].Type() == diag.ACCESS {
```
Comment (erezrokah) [inline client/client.go:446]: @disq is this what you had in mind?

Is there a better way to access the `Type` and create a new diag?

Comment (disq) [reply client/client.go:446]: Not at all, just edit `classifyError` to match the returned err to a diag of proper kind

Comment (disq) [reply client/client.go:446]: so:
step 1- determine how to detect these user errors in `classifyError`, using `smithy.APIError.ErrorCode()` or other ways
step 2- create the diag with `diag.WithType(diag.USER)` for these errors inside `classifyError`
step 3- call `classifyError` in all places to turn `error`s into `diag.Diagnostics`

classifyError takes a `fallbackType` argument, so if the error wasn't classified it creates a diag with the given fallback type.

Comment (erezrokah) [reply client/client.go:446]: Thanks for the detailed explanation @disq. My concern is that we might get this error not during initial setup which could be classified differently. Do we know if it's possible to have the same error as a transient one that we can recover from by retrying?

File: client/errors.go [hunk 4]
```diff
@@ -22,13 +22,13 @@ func classifyError(err error, fallbackType diag.Type, accounts []Account, opts .
 	var ae smithy.APIError
 	if errors.As(err, &ae) {
 		switch ae.ErrorCode() {
-		case "AccessDenied", "AccessDeniedException", "UnauthorizedOperation", "AuthorizationError", "OptInRequired", "SubscriptionRequiredException", "InvalidClientTokenId":
+		case "AccessDenied", "AccessDeniedException", "UnauthorizedOperation", "AuthorizationError", "OptInRequired", "SubscriptionRequiredException", "InvalidClientTokenId", "ExpiredToken":
```
Comment (erezrokah) [inline client/errors.go:25]: `ExpiredToken` happens when the SSO login token from our playground account expires

File: client/client.go (hunk 3 above)
Comment (disq) [reply client/client.go]: Not sure, possibly (transient error) but we don't retry explicitly, leave it to the aws sdk (which is more than capable)

**Review by disq — APPROVED** [review]

**Review by roneli — APPROVED** [review]
LGTM

**Pull request merged**